src/function/scalar/list/list_select.cpp

```diff
@@ -0,0 +1,103 @@
+#include "duckdb/common/types/data_chunk.hpp"
+#include "duckdb/function/scalar/nested_functions.hpp"
+#include "duckdb/planner/expression_binder.hpp"
+#include "duckdb/planner/expression/bound_function_expression.hpp"
+#include "duckdb/planner/expression/bound_parameter_expression.hpp"
+
+namespace duckdb {
+
+static void ListSelectFunction(DataChunk &args, ExpressionState &state, Vector &result) {
+	Vector &list = args.data[0];
+	Vector &selection_list = args.data[1];
+	idx_t count = args.size();
+
+	ListVector::Reserve(result, ListVector::GetListSize(selection_list));
+
+	bool all_constant = true;
+	for (idx_t i = 0; i < args.ColumnCount(); i++) {
+		auto curr = args.data[i];
+		// check if all arguments are constants, if so the result is also a constant vector
+		if (curr.GetVectorType() != VectorType::CONSTANT_VECTOR) {
+			all_constant = false;
+		}
+	}
+
+	list_entry_t *result_data;
+	result_data = FlatVector::GetData<list_entry_t>(result);
+	auto &result_entry = ListVector::GetEntry(result);
+
+	UnifiedVectorFormat sel_list;
+	selection_list.ToUnifiedFormat(count, sel_list);
+	auto selection_lists = UnifiedVectorFormat::GetData<list_entry_t>(sel_list);
+	auto &sel_entry = ListVector::GetEntry(selection_list);
+
+	UnifiedVectorFormat inp_list;
+	list.ToUnifiedFormat(count, inp_list);
+	auto input_lists = UnifiedVectorFormat::GetData<list_entry_t>(inp_list);
+	auto &input_entry = ListVector::GetEntry(list);
+
+	idx_t offset = 0;
+	for (idx_t j = 0; j < count; j++) {
+		// Get length and offset of selection list for current output row
+		auto sel_list_idx = sel_list.sel->get_index(j);
+		idx_t sel_len = 0;
+		idx_t sel_off = 0;
+		if (sel_list.validity.RowIsValid(sel_list_idx)) {
+			sel_len = selection_lists[sel_list_idx].length;
+			sel_off = selection_lists[sel_list_idx].offset;
+		}
+
+		// Get length and offset of input list for current output row
+		auto inp_list_idx = inp_list.sel->get_index(j);
+		idx_t inp_len = 0;
+		idx_t inp_off = 0;
+		if (inp_list.validity.RowIsValid(inp_list_idx)) {
+			inp_len = input_lists[inp_list_idx].length;
+			inp_off = input_lists[inp_list_idx].offset;
+		}
+
+		// Set all selected values in the result
+		for (idx_t i = 0; i < sel_len; i++) {
+			idx_t sel_idx = sel_entry.GetValue(sel_off + i).GetValue<idx_t>();
+			Value sel_val = sel_idx < inp_len ? input_entry.GetValue(inp_off + sel_idx) : Value(LogicalType::SQLNULL);
+			result_entry.SetValue(offset + i, sel_val);
+		}
+
+		result_data[j].length = sel_len;
+		result_data[j].offset = offset;
+		offset += sel_len;
+	}
```

Comment:
Some thoughts on this.
- We need to set the validity of the result vector correctly. I.e., the following indexes in the selection list should result in a `NULL` field in the result vector: negative index, index exceeding the list length, `NULL`.

Also, setting each value iteratively here should not be necessary. Optionally, I strongly prefer an implementation using a selection vector that slices the input, as that should yield better performance for nested types and `VARCHARs`, and similar performance for numeric types. If you decide not to implement this, then please add a `FIXME`, so that we can pick this up at a later point.

You can create a selection vector that has the same size as the selection list. Then, you iterate over the two input lists. You set the validity to invalid for all negative indexes, indexes exceeding the list size, and `NULLs`. Then, you set the selection vector for all remaining values: `sel_vector.set_index(offset + i, sel_val);`. You might have to set the selection vector to a dummy index for `NULLs`, I am not sure. After the loop, you can slice the input list and store the result in the result vector. `Slice` will create a dictionary vector.

Nit: maybe use `child_idx` instead of `i`.

Reply:
In terms of validity, are invalid values then as 0 represented? Is there a good way that invalid fields are NULLs instead?